RFC: Return Type Notation (#3654)
rust-lang/rfcs: text/0000-return-type-notation.md

# Return Type Notation in Bounds and Where-clauses

- Feature Name: `return_type_notation`
- Start Date: 2024-06-04
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Return type notation (RTN) gives a way to reference or bound the type returned by a trait method.

> The new bounds look like `T: Trait<method(..): Send>` or `T::method(..): Send`.

Reviewer: I think it could be cleaner to leverage [`FnOnce::Output`](https://doc.rust-lang.org/std/ops/trait.FnOnce.html#associatedtype.Output) somehow, which then wouldn't require any new syntax. E.g. `T::method::Output: Send` desguaring to something like `<<T as Trait>::method as FnOnce(..) -> _>::Output: Send`. 

Unfortunately similar patterns hit "the precise format of `Fn`-family traits' type parameters is subject to change". Not sure how feasible it would be to consider at least the `Output` associated type stable, but I would _love_ to be able to use this elsewhere too.

Author: We explored this option. This exploration is covered under the FAQ ["What other syntax options were considered"](https://github.com/nikomatsakis/rfcs/blob/rtn/text/0000-return-type-notation.md#what-other-syntax-options-were-considered) which, for this particular proposal, bottomed out in a [design like this one](https://smallcultfollowing.com/babysteps/blog/2023/06/12/higher-ranked-projections-send-bound-problem-part-4/). One major drawback noted in the RFC:

> This reuses associated types but, as both the function and future traits define an `Output` associated type, raises the potential for confusion about whether this notation means "the future that gets returned" or "the result of the future".

We ultimately decided against this -- but we also noted that it could be added in the future, in which case `(..)` could be seen as a shorthand for `::Output`, essentially.

>
>The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.

## Background and running examples

### The `Widgets` trait

Throughout this section we will make use of the `Widgets` trait as a simple running example.

```rust
trait Widgets {
    fn widgets(&self) -> impl Iterator<Item = Widget>;
}
```

### Background: desugaring to associated types

Per [RFC 3425][], the return-position `impl Trait` types that appear in `Widgets` and `Log` are desugared by the compiler into generic associated types, roughly as follows:

[RFC 3425]: https://github.com/rust-lang/rfcs/pull/3425

```rust
trait Widgets { // desugared
    type $Widgets<'a>: Iterator<Item = u32>;
    fn widgets(&self) -> Self::$Widgets<'_>;
}
```

These desugarings are not exposed to users, so the associated types `$Widgets` and `$Log` are not directly nameable,
but we will use it to define the semantics of Return Type Notation.

## Grammar

### Return type notation

Return Type Notation extends the type grammar roughly as follows,
where `?` indicates an optional nonterminal and `,*` indicates a comma

> separated list. These changes permit `where T::method(): Send`.

Reviewer 2: ```suggestion
separated list. These changes permit `where T::method(..): Send`.
```

* `<T as Widgets>::widgets(..): Send`
    * `where for<'a> <T as Widgets>::$Widgets<'a>: Send`
* `T: Widgets<widgets(..): Send`
    * Equivalent to `where T: Widgets<for<'a> $Widgets<'a>: Send>`
* `impl Widgets<widgets(..): Send>`
    * `impl for<'a> Widgets<$Widgets<'a>: Send>`
* `dyn Widgets<widgets(..): Send`
    * `dyn for<'a> Widgets<$Widgets<'a>: Send>`
    * But note that async fn and RPITIT are not yet dyn-safe; this is forward looking.

While all of these examples are using lifetimes, there is ongoing work to support higher-ranked trait bounds that are generic over types, and the expectation is that RTN will be extended to work over generic types and constants when possible.

### How this is implemented

The examples above illustrate the semantics but do not make clear how RTN can be implemented in the compiler. A RTN bound like `widgets(..)` is implemented internally via unification. To keep the RFC focused on how RTN feels to users, we defer a detailed description to reference material and a future stabilization report.

## RTN only applies to AFIT and RPITIT methods

> Although conceptually RTN could be used for any trait method, we choose to limits its use to `async fn` and other methods that directly return an `-> impl Future`. This limitation can be lifted in the future as we gain more experience.

Reviewer 3: ```suggestion
Although conceptually RTN could be used for any trait method, we choose to limits its use to `async fn` and other methods that directly return an `-> impl Trait`. This limitation can be lifted in the future as we gain more experience.
```

* "FnDotDot" and friends: `D: Database<fn items(..): DoubleEndedIterator>`
    * This notation was deemed too close to `fn` pointer types, particularly in stand-alone where-clauses.

## Why not use `typeof`, isn't that more general?

The compiler currently supports a `typeof` operation as an experimental feature (never RFC'd). The idea is that `typeof <expr>` type-checks `expr` and evaluates to the result of that expression. Therefore `typeof 22_i32` would be equivalent to `i32`, and `typeof x` would be equivalent to whatever the type of `x` is in that context (or an error if there is no identifier `x` in scope). 

It might appear that `typeof`  can be used in a similar way to RTN, but in fact it is significantly more complex. Consider our first example, the `HealthCheck` trait:

```rust
trait HealthCheck {
    async fn check(&mut self, server: Server);
}
```

and a function bounding it

```rust
fn start_health_check<H>(health_check: H, server: Server)
where
    H: HealthCheck + Send + 'static,

>     H::check(): Send, // 👈 How would we write this with `typeof`?

Reviewer 2: ```suggestion
    H::check(..): Send, // 👈 How would we write this with `typeof`?
```

## Why not create an associated type that represents the return type?

Early on in our design work, we expected to simply create an associated type within the trait to represent the return type. For example this trait:

```rust
trait Factory {
    fn widgets(&self) -> impl Iterator<Item = Widget>;
}
```

might have been desugared as follows:

```rust
trait Factory {
    type widgets<'a>: Iterator<Item = Widget>; // 👈 implicitly introduced
    fn widgets(&self) -> Self::widgets<'_>;
}
```

This would mean that users could write a bound on `F::widgets` to bound the return type of `widgets`

```rust
fn use_factory<F>()
where
    F: Factory,
    for<'a> F::widgets<'a>: Send,
{}
```

We encountered a number of problems with this design.

> ### If the name is implicit, what name should we use?

Reviewer 2: you could also use one of those `#` namespace such as `fn#widgets` or `return#widgets` or `result_of#widgets`

Author: yes, I'll add that as a possibility

The most impmediate problem with this proposal was trying to decide what name to use.

Using `Widgets` (capitalized) feels arbitrary and there is no precedent within Rust for automatically creating names with different case conventions in this way.

Using the same name as the method (`widgets`) results in an associated type that does not follow Rust's naming conventions.
It also introduces the potential for a shadowing conflict as today it is allowed to have methods and associated types with the same name:

```rust
trait Example {
    type method;
    fn method();
}
```

### Why not use an explicit name?

To address the challenge of an implicit name, we could allow people to explicitly annotate a name:

```rust
trait Factory {

>     #[associated_return_type(Widget)]

Reviewer 2: ```suggestion
    #[associated_return_type(Widgets)]
```

The `T::method(..): Send` notation we've been using so far means
"the return type of `method(..)` is `Send`, no matter what arguments you provide".
We could extend this notation to permit specifying the argument types explicitly.
For example, consider the `capture` method below, which takes a parameter of type `input`:

```rust
trait Capture {
    async fn capture<T>(&mut self, input: T) -> Option<T>;
}
```

and now consider a function that invokes `capture` with an `i32` value:

```rust
async fn capture_i32<C: Capture>(mut c: C) {
    c.capture(22_i32);
}
```

Now imagine we wanted to invoke `capture` on another thread,
and hence we need a where-clause indicating that the future
returned by `capture` will be `Send`:

```rust
async fn capture_i32<C: Capture + Send + 'static>(mut c: C)
where
    /* where-clause for C::check() needed here! */
{
    workstealing_runtime::spawn(async move {
        c.capture(22_i32);
    })
}
```

There are multiple ways we could write this where-clause, varying in their specificity...

* `where C::capture(..): Send` -- this indicates that `C::capture()` will return a `Send` value for any possible set of parameters

> * `where C::capture(&mut C, i32): Send` -- this indicates that `C::capture()` will return a `Send` value when invoked specifically on a `&mut C` (for the `self` parameter) and an `i32`

Reviewer 2: wouldn't this bring back the ambiguity with `Fn()`? for example when we have `fn capture() -> impl Future`

Reviewer 4: Indeed, it would. `fn f() where C::capture(&mut C, i32): Send {}` is already syntactically legal *today*. The parentheses denote *parenthesized type arguments* which may be attached to arbitary type paths (and it also comes with turbofish for expression and pattern paths: `C::capture::(&mut C, i32)`).

Author: It is syntactically legal, yes, but not meaningful semantically.

Reviewer 4: Semantically as well if `C` is a module but yeah, I don't think it can actually meaningfully clash if `C` is a type parameter.

```rs
//@ edition: 2015
//@ check-pass
mod C { pub use std::ops::Fn as capture; }
fn f() where C::capture(): 'static {}
```

* `where for<'a> C::capture(&'a mut C, i32): Send` -- same as the previous rule, but with the higher-ranked `'a` written explicitly

> * `where C::capture::<i32>(): Send` -- this indicates that `C::capture()` will return a `Send` value for any possible set of parameters, but with its `T` parameter set explicitly to `i32`

Reviewer 2: ```suggestion
* `where C::capture::<i32>(..): Send` -- this indicates that `C::capture()` will return a `Send` value for any possible set of parameters, but with its `T` parameter set explicitly to `i32`
```

    //   value.
    await_future(load_future);
}
    
fn await_future<D: DataFactory>(load_future: D::load(..)) -> Data {
    //                                       -------
    //                      As above, expands to `D::load(&'_ D)`, which
    //                      means "for some `_`".
    argument.await
}
```

The most useful place to use RTN, however, is likely struct fields, and in that location we do not accept `'_`.
We would therefore have to support specifying the types of arguments in RTN.
That would enable writing structs that wrap the future returned via some trait method:

```rust
struct Wrap<'a, D: DataFactory> {
    load_future: D::load(&'a D), // the future returned by `D::load`.
}
```

## Dyn support

We expect to  make traits with async functions and RPITIT dyn safe in the future. One benefit of the RTN design is that it continues to hide the presence and precise value of the associated types that define the return value of an async function. This means that given `HealthCheck`, we can later define the type of the future `<dyn HealthCheck>::check(..)` to be anything.

## Naming the zero-sized types for a method

> Every function and method `f` in Rust has a corresponding zero-sized type that uniquely identifies `f`. The RTN notation `T::check(..)` refers to the return value of `check`; conceivably `T::check` (without the parens) could be used to refer the type of `check` itself. In this case, `T::check()` can be thought of as shorthand for `<T::check as Fn<_>>::Output`.

Reviewer 2: ```suggestion
Every function and method `f` in Rust has a corresponding zero-sized type that uniquely identifies `f`. The RTN notation `T::check(..)` refers to the return value of `check`; conceivably `T::check` (without the parens) could be used to refer the type of `check` itself. In this case, `T::check(..)` can be thought of as shorthand for `<T::check as Fn<_>>::Output`.
```

# Return Type Notation in Bounds and Where-clauses

- Feature Name: `return_type_notation`
- Start Date: 2024-06-04
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Return type notation (RTN) gives a way to reference or bound the type returned by a trait method. The new bounds look like `T: Trait<method(..): Send>` or `T::method(..): Send`. The primary use case is to add bounds such as `Send` to the futures returned by `async fn`s in traits and `-> impl Future` functions, but they work for any trait function defined with return-position impl trait (e.g., `where T: Factory<widgets(..): DoubleEndedIterator>` would also be valid).

> This RFC proposes a new kind of type written `<T as Trait>::method(..)` (or `T::method(..)` for short). RTN refers to "the type returned by invoking `method` on `T`". 

Reviewer 5: Technically, in the general case the two forms here must represent a set of types, not just a single type, right?

That is, if the assigned return type is derived from the argument types, then it is not correct to say "the type returned", right? (I'm referring to the kind of issue described at the end of the RFC in the section entitled "Specifying the values for argument types")

Reviewer 5: (I'd be fine if this were handled as a footnote.)

Author: Correct.

* Referencing `T: SendMyTrait` is shorter than using RTN if there are multiple functions
    * (compare to `T: Send + Mytrait<method1(..): Send, method2(..): Send>`)
* Referencing `T: SendMyTrait` is more forwards compatible:
    * If you add a new method to your trait (with a default impl), all users of the send alias will be able to call this new method. Users that have named individual methods will not (on the flip side)

But defining a "send alias" in this way comes with obligations for you:

* If you add a new default method to your trait, it must be "Send-preserving" (meaning that it will be `Send` if other functions return `Send` futures).
    * *Why?* If there is an existing function that requires `T: SendMyTrait` for some type `T`, then this must remain true even when `MyTrait` grows a new (defaulted) method, or else you will have broken your downstream clients.
    * On the flip side, if you don't define an alias, you can add new defaulted methods that are not Send. This won't break downstream crates but neither will they be able to use them.

### Using async traits

When using a trait `MyTrait` that defines a sendable alias `SendMyTrait`...

> * Implement `MyTrait` directly. Your type will implement `SendMyTrait` automatically if appropriate.

Reviewer 6: Interesting.. I had to think about whether I agree with this and started to argue against it, but I think it's probably the right call.

Implementing `SendMyTrait` directly can be nice for concrete types, because makes it your contracts more explicit. Of course this is a special case of auto trait leakage, but here we have the opportunity to mitigate its downsides with best practices. But a big problem with this take is that the converse is not true; implementing `MyTrait` does not remove any semver hazards from your code, it only hides them. I would not want to lead users into a false sense of security here.

The other problem is that the practice of implementing `SendMyTrait` doesn't extend to generic types. You could argue that the resulting need to add a `where T: Send` or `where T: SendMyTrait` to `impl<T> SendMyTrait for Struct<T>` might cause an implementer to think twice about whether their impl needs to be so constrained. But then the key question becomes whether they _know_ they can write `impl<T> MyTrait for Struct<T>` and still have `SendMyTrait` implemented when possible. In the case where you always implement `MyTrait` directly, this quickly becomes obvious from experience.

With all that said, I might personally choose to implement `SendMyTrait` directly in cases where it's important.. but that would be going against the general best practices laid out here (which I agree with), and is probably best left for the cases where I would otherwise bother to write a static assertion.

Note that this is connected to the convention of whether to use `MyTrait`/`SendMyTrait` or `LocalMyTrait`/`MyTrait`. In the latter case it would seem we are encouraging impls to use the send variant by default and to "opt in" to more genericity / fewer guarantees.

Author: Good points. I think when I wrote this, I was also thinking about the fact that you cannot implement trait aliases today. I think we should expedite work on changing that, in which case I suspect that the best advice would be to implement the Send variant if that's what you always want.

Author: I'm not so concerned about people learning that they can implement the 'non-send' variant and still get the benefits. They'll get that in time, I think, I don't think they have to learn it right away. Using the send variant will get you faster and better error messages.

# Unresolved questions
[unresolved questions]: #unresolved-questions

>- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?

## Does stabilizing `T::foo(..)` notation as a standalone type create a confusing inconsistency with `-> ()` shorthand?

Unlike a regular associated type, this RFC does not allow a trait bound that specifies the return type of a method, only the ability to put bounds on that return type.
rpjohnst suggested that we may wish to support a syntax like `T: Trait<method(..) -> T>`, perhaps in conjunction with specified argument types.
They further pointed out that permitting `T::method(..)` as a standalone type could be seen as inconsistent, given that `fn foo()` is normally shorthand for `-> ()`.
However, *not* supporting `T::method(..)` as a standalone type could also be seen as inconsistent, since normally `T: Trait<Bar: Send>` and `T::Bar: Send` are equivalent.

> Prior to stabilizing the "send position" syntax, we should be sure we are comfortable with this.

Reviewer 6: ```suggestion
Prior to stabilizing the "associated type position" syntax, we should be sure we are comfortable with this.
```

I think this is what you meant?
